tinygrad/codegen/linearizer.py

```diff
@@ -428,29 +436,28 @@ def linearize(self):
     return self
 
   def ast_parse(self, x:LazyOp, acc: List[UOp], offs:Optional[List[int]], loaded_buffers:Dict[Union[MemBuffer, ConstBuffer, LocalBuffer], List[UOp]],\
-                do_reduce=False, loop_ctx=tuple(), cache=None, insert_before:Optional[UOp]=None) -> List[UOp]:
+                reduced_ops: Dict[LazyOp, List[UOp]]={}, do_reduce=False, loop_ctx=tuple(), cache=None) -> List[UOp]:
     if cache is None: cache = {}
     if x in cache: return cache[x]
     if x.op in BufferOps: return loaded_buffers[x.arg]
-    if x.op is UnaryOps.CAST: return [self.uops.add(UOps.BITCAST if x.arg[1] else UOps.CAST, self.get_base_dtype(x.arg[0]), (u,), x.arg[0], \
-                  insert_before=insert_before) for u in self.ast_parse(x.src[0], acc, offs, loaded_buffers, insert_before=insert_before)]
+    if x.op is UnaryOps.CAST: return [self.uops.add(UOps.BITCAST if x.arg[1] else UOps.CAST, self.get_base_dtype(x.arg[0]), (u,), x.arg[0]) \
+                                       for u in self.ast_parse(x.src[0], acc, offs, loaded_buffers, reduced_ops)]
     if x.op in ReduceOps and not do_reduce:
-      assert offs is None, "not available if we aren't doing reduce"
-      return acc
+      return [reduced_ops[x][i] for i in offs] if offs else reduced_ops[x]
```

Comment:
i think offs just removes unrolled shapes from iteration, so just returning reduced_ops[x] here is correct.

Reply:
maybe I goofed something elsewhere and this just covers it up, but this was so that, if there's an intermediate calculation between two reduceops, it gets brought back to the unrolled shape

ex: an unrolled standard deviation looks like this without of the offs
```
__kernel void r_4_32_8n4(__global float* data0, const __global float* data1) {
  int gidx0 = get_group_id(0); /* 4 */
  int lidx1 = get_local_id(0); /* 32 */
  float acc0 = 0.0f;
  int alu0 = ((gidx0*256)+(lidx1*8));
  float val0 = data1[alu0];
  float val1 = data1[alu0+1];
  float val2 = data1[alu0+2];
  float val3 = data1[alu0+3];
  float val4 = data1[alu0+4];
  float val5 = data1[alu0+5];
  float val6 = data1[alu0+6];
  float val7 = data1[alu0+7];
  float acc1 = 0.0f;
  data0[(gidx0*32)+lidx1] = (0.015625f*((val0-((val7+val6+val5+val4+val3+val2+val1+val0+acc0)*0.015625f))+acc1));
}
```

when what we want is this:
```
__kernel void r_4_32_8n4(__global float* data0, const __global float* data1) {
  int gidx0 = get_group_id(0); /* 4 */
  int lidx1 = get_local_id(0); /* 32 */
  float acc0 = 0.0f;
  int alu0 = ((gidx0*256)+(lidx1*8));
  float val0 = data1[alu0];
  float val1 = data1[alu0+1];
  float val2 = data1[alu0+2];
  float val3 = data1[alu0+3];
  float val4 = data1[alu0+4];
  float val5 = data1[alu0+5];
  float val6 = data1[alu0+6];
  float val7 = data1[alu0+7];
  float acc1 = 0.0f;
  float alu1 = ((val7+val6+val5+val4+val3+val2+val1+val0+acc0)*0.015625f);
  data0[(gidx0*32)+lidx1] = (0.015625f*((val7-alu1)+(val6-alu1)+(val5-alu1)+(val4-alu1)+(val3-alu1)+(val2-alu1)+(val1-alu1)+(val0-alu1)+acc1));
}
```

Reply:
oh, i see. makes sense!